llama.cpp

```diff
@@ -1452,8 +1452,8 @@ static bool llama_model_quantize_internal(const std::string & fname_inp, const s
             fout.write((char *) &len, sizeof(len));
 
             word.resize(len);
-            finp.read ((char *) word.data(), len);
-            fout.write((char *) word.data(), len);
+            finp.read (&word[0], len);
+            fout.write(&word[0], len);
```

Comment:
I don't think this is correct, as .data() is used everywhere else for std::strings. If it is, then it should be changed everywhere else too?

To be honest this whole class could use a overhaul, I had a quick look at this today and there is a lot of perfomance to gain by changing std::strings and std::vector<char>'s to contiguous memory blocks with primitive types. Especially the parts where the temporary buffers are being .resize()'d constantly is a huge performance penalty especially when going back and forth to being smaller and larger.

Reply:
> `.data()` is presumably used as `const` everywhere else?
Nope, they are used as char* as buffers to write into. So definitely not const. If thats UB, that doesn't sound good. In any case the larger underlying problem is using those constructs at all when they are actually being used as regular memory blocks for the file I/O operations. The PR #216 touches on the issue.

However, since everything works now and it's not segfaulting anymore like it was a while back, I think it's better left untouched and later do a complete overhaul to move the file I/O to primitive types and regular CRT fopen() operations. Mixing STL and raw pointers can easily lead to hard-to-debug memory management and access violation problems but it does work now so I wouldn't change it. :smile: